internal/querynode/shard_cluster.go

```diff
@@ -381,20 +400,166 @@ func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+		}
+		// reference count ++
+		segment.inUse++
 		result[segment.nodeID] = append(result[segment.nodeID], segment.segmentID)
 	}
 	return result
 }
 
+// inHandoffOffline checks whether segment is pending handoff offline list
+// Note that sc.mut Lock is assumed to be hold outside of this function!
+func (sc *ShardCluster) inHandoffOffline(segmentID int64) bool {
+	for _, handoff := range sc.handoffs {
+		for _, offlineSegment := range handoff.OfflineSegments {
+			if segmentID == offlineSegment.GetSegmentID() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// finishUsage decreases the inUse count of provided segments
+func (sc *ShardCluster) finishUsage(allocs map[int64][]int64) {
+	defer func() {
+		sc.rcCond.L.Lock()
+		sc.rcCond.Broadcast()
+		sc.rcCond.L.Unlock()
+	}()
+	sc.mut.Lock()
+	defer sc.mut.Unlock()
+	for _, segments := range allocs {
+		for _, segmentID := range segments {
+			segment, ok := sc.segments[segmentID]
+			if !ok || segment == nil {
+				// this shall not happen, since removing segment without decreasing rc to zero is illegal
+				log.Error("finishUsage with non-existing segment", zap.Int64("collectionID", sc.collectionID), zap.Int64("replicaID", sc.replicaID), zap.String("vchannel", sc.vchannelName), zap.Int64("segmentID", segmentID))
+				continue
+			}
+			// decrease the reference count
+			segment.inUse--
+		}
+	}
+}
+
+// handoffSegments processes the handoff/load balance segments update procedure.
+func (sc *ShardCluster) HandoffSegments(info *querypb.SegmentChangeInfo) error {
+	// wait for all OnlineSegment is loaded
+	onlineSegments := make([]int64, 0, len(info.OnlineSegments))
+	for _, seg := range info.OnlineSegments {
+		// filter out segments not maintained in this cluster
+		if seg.GetCollectionID() != sc.collectionID || seg.GetDmChannel() != sc.vchannelName {
+			continue
+		}
+		onlineSegments = append(onlineSegments, seg.GetSegmentID())
+	}
+	sc.waitSegmentsOnline(onlineSegments)
+
+	// add segmentChangeInfo to pending list
+	token := sc.appendHandoff(info)
+
+	// wait for all OfflineSegments is not in use
+	offlineSegments := make([]int64, 0, len(info.OfflineSegments))
+	for _, seg := range info.OfflineSegments {
+		offlineSegments = append(offlineSegments, seg.GetSegmentID())
+	}
+	sc.waitSegmentsNotInUse(offlineSegments)
+	// remove offline segments record
+	for _, seg := range info.OfflineSegments {
+		// filter out segments not maintained in this cluster
+		if seg.GetCollectionID() != sc.collectionID || seg.GetDmChannel() != sc.vchannelName {
+			continue
+		}
+		sc.removeSegment(segmentEvent{segmentID: seg.GetSegmentID(), nodeID: seg.GetNodeID()})
+	}
+
+	// finish handoff and remove it from pending list
+	sc.finishHandoff(token)
+
+	return nil
+}
+
+// appendHandoff adds the change info into pending list and returns the token.
+func (sc *ShardCluster) appendHandoff(info *querypb.SegmentChangeInfo) int32 {
+	sc.mut.Lock()
+	defer sc.mut.Unlock()
+
+	token := sc.nextToken.Add(1)
```

Comment:
If `nextToken` means the next token of `SegmentChangeInfo` to be used, I think we should return the value of the token before incrementing.

Reply:
Fixed